bigquery_etl/publish_json.py

```diff
@@ -0,0 +1,148 @@
+        self.parameter = parameter
+        self.client = client
+        self.storage_client = storage_client
+        self.temp_table = None
+        self.date = None
+
+        self.metadata = Metadata.of_sql_file(self.query_file)
+
+        if self.parameter:
+            date_search = re.search(SUBMISSION_DATE_RE, self.parameter)
+
+            if date_search:
+                self.date = date_search.group(1)
+
+        query_file_re = re.search(QUERY_FILE_RE, self.query_file)
+        if query_file_re:
+            self.dataset = query_file_re.group(1)
+            self.table = query_file_re.group(2)
+            self.version = query_file_re.group(3)
+        else:
+            print("Invalid file naming format: {}", self.query_file)
+            sys.exit(1)
+
+    def __exit__(self):
+        """Delete temporary tables."""
+        if self.temp_table:
+            self.client.delete_table(self.temp_table)
+
+    def publish_json(self):
+        """Publish query results as JSON to GCP Storage bucket."""
+        if self.metadata.is_incremental():
+            if self.date is None:
+                print("Cannot publish JSON. submission_date missing in parameter.")
+                sys.exit(1)
+
+            # if it is an incremental query, then the query result needs to be
+            # written to a temporary table to get exported as JSON
+            self._write_results_to_temp_table()
+            self._publish_table_as_json(self.temp_table)
+        else:
+            # for non-incremental queries, the entire destination table is exported
+            result_table = f"{self.dataset}.{self.table}_{self.version}"
+            self._publish_table_as_json(result_table)
+
+    def _publish_table_as_json(self, result_table):
+        """Export the `result_table` data as JSON to Cloud Storage."""
+        prefix = (
+            f"api/{self.api_version}/tables/{self.dataset}/"
+            f"{self.table}/{self.version}/files/"
+        )
+
+        if self.date is not None:
+            # if date exists, then query is incremental and newest results are exported
+            prefix += f"{self.date}/"
+
+        table_ref = self.client.get_table(result_table)
+
+        job_config = bigquery.ExtractJobConfig()
+        job_config.destination_format = "NEWLINE_DELIMITED_JSON"
+
+        # "*" makes sure that files larger than 1GB get split up into JSON files
+        destination_uri = f"gs://{self.target_bucket}/" + prefix + "*.json"
+        extract_job = self.client.extract_table(
+            table_ref, destination_uri, location="US", job_config=job_config
+        )
+        extract_job.result()
+
+        self._gcp_convert_ndjson_to_json(prefix)
+
+    def _gcp_convert_ndjson_to_json(self, gcp_path):
+        """Convert ndjson files on GCP to json files."""
+        blobs = self.storage_client.list_blobs(self.target_bucket, prefix=gcp_path)
+        bucket = self.storage_client.bucket(self.target_bucket)
+
+        for blob in blobs:
```

Comment:
We need to ensure this is an atomic operation, such that if we fail at any point during any of these blobs, it looks as if no data is processed. In that case we should backup the data to a saved location, write the data to a staging location, overwrite the prod data with that staging data, then delete the backup data. With this scenario only time we may see an atomicity fail is during overwriting the prod data with stage data, so we would have to check for that kind of failure and either write the backups (in case of non-incremental queries) or delete the data entirely for that day (in case of incremental queries). Let me know if you want to chat about this.